File: misc/aooswd/ex/strategy/cpp/best_fit_allocation.cpp
#include "best_fit_allocation.h"

void BestFitAllocation::allocate (list<MemoryBlock>& l, int size)
{
  success = false;
  if (!l.empty())
  {
    list<MemoryBlock>::iterator position = l.begin();
    list<MemoryBlock>::iterator best;
    while (position != l.end())
    {
      if (position -> size() >= size)
	if (!success)
	{
	  success = true;
	  best = position;
	}
	else if (position -> size() < best -> size())
	  best = position;
      position++;
    }
    if (success)
      allocateBlock (l, size, best);
  }
}

File: misc/aooswd/A6/2/BestFitAllocation.java
/*
 * BestFitAllocation.java
 *
 * Created on January 23, 2003, 11:25 AM
 */

import java.util.List;

/**
 *
 * @author  dgayler
 */
public class BestFitAllocation extends MemoryAllocation
{
    
    /** Creates a new instance of BestFitAllocation */
    public BestFitAllocation()
    {
    }
      
    public void allocate(List l, int size) throws AddressOrderException
    {
        success = false;
        if (!l.isEmpty())
        {
            int position = 0;
            int best = 0;
            while (position < l.size())
            {
                if (((MemoryBlock)l.get(position)).size() >= size)
                    if (!success)
                    {
                        success = true;
                        best = position;
                    }
                    else if (((MemoryBlock)l.get(position)).size() < ((MemoryBlock)l.get(best)).size())
                        best = position;
                position++;
            }
            if (success)
                allocateBlock (l, size, best);
        }
    }

}
